src/manifolds/StiefelEuclideanMetric.jl

```diff
@@ -0,0 +1,144 @@
+default_metric_dispatch(::Stiefel, ::EuclideanMetric) = Val(true)
+
+function decorator_transparent_dispatch(::typeof(distance), ::Stiefel, args...)
+    return Val(:intransparent)
+end
+
+@doc raw"""
+    exp(M::Stiefel, p, X)
+
+Compute the exponential map on the [`Stiefel`](@ref)`{n,k,𝔽}`() manifold `M`
+emanating from `p` in tangent direction `X`.
+
+````math
+\exp_p X = \begin{pmatrix}
+   p\\X
+ \end{pmatrix}
+ \operatorname{Exp}
+ \left(
+ \begin{pmatrix} p^{\mathrm{H}}X & - X^{\mathrm{H}}X\\
+ I_n & p^{\mathrm{H}}X\end{pmatrix}
+ \right)
+\begin{pmatrix}  \exp( -p^{\mathrm{H}}X) \\ 0_n\end{pmatrix},
+````
+
+where $\operatorname{Exp}$ denotes matrix exponential,
+$\cdot^{\mathrm{H}}$ denotes the complex conjugate transpose or Hermitian, and $I_k$ and
+$0_k$ are the identity matrix and the zero matrix of dimension $k × k$, respectively.
+"""
+exp(::Stiefel, ::Any...)
+
+function exp!(::Stiefel{n,k}, q, p, X) where {n,k}
+    return copyto!(
+        q,
+        [p X] *
+        exp([p'X -X'*X; one(zeros(eltype(p), k, k)) p'*X]) *
+        [exp(-p'X); zeros(eltype(p), k, k)],
+    )
```

Comment:
This is simpler and avoids a redundant product
```suggestion
    A = p'*X
    return copyto!(q, [p X] * exp([A -X'*X; I p'*X]) * [exp(-A); 0*I])
```